<a href="https://colab.research.google.com/github/Devesh-Patodia/DL-J038/blob/labwork/Lab_6(ii)_Keras_Activations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of activations

Activations can either be used through an Activation layer, or through the activation argument supported by all forward layers:



```
from keras.layers import Activation, Dense
model.add(Dense(64))
model.add(Activation('tanh'))
```



This is equivalent to:

```
model.add(Dense(64, activation='tanh'))
```

You can also pass an element-wise TensorFlow/Theano/CNTK function as an activation:

```
from keras import backend as K
model.add(Dense(64, activation=K.tanh))
```

#elu

keras.activations.elu(x, alpha=1.0)

Exponential linear unit.

##Arguments

1. x: Input tensor.
2. alpha: A scalar, slope of negative section.

Returns

1. The exponential linear activation: x if x > 0 and alpha * (exp(x)-1) if x < 0.

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 12s 192us/step - loss: 0.2837 - acc: 0.9132 - val_loss: 0.1978 - val_acc: 0.9393
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1214 - acc: 0.9625 - val_loss: 0.1060 - val_acc: 0.9666
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0837 - acc: 0.9737 - val_loss: 0.0874 - val_acc: 0.9734
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0631 - acc: 0.9801 - val_loss: 0.0883 - val_acc: 0.9736
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0494 - acc: 0.9838 - val_loss: 0.0884 - val_acc: 0.9741
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0393 - acc

#softmax

keras.activations.softmax(x, axis=-1)

##Softmax activation function.

##Arguments

1. x: Input tensor.
2. axis: Integer, axis along which the softmax normalization is applied.

##Returns

Tensor, output of softmax transformation.

##Raises

ValueError: In case dim(x) == 1.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softmax', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softmax'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 32us/step - loss: 2.2688 - acc: 0.1762 - val_loss: 2.1954 - val_acc: 0.5574
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 2.0404 - acc: 0.7998 - val_loss: 1.8505 - val_acc: 0.8391
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.6371 - acc: 0.8393 - val_loss: 1.4163 - val_acc: 0.8450
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.2142 - acc: 0.8449 - val_loss: 1.0209 - val_acc: 0.8476
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.8645 - acc: 0.8485 - val_loss: 0.7276 - val_acc: 0.8503
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.6332 - acc: 0.8513 - val_loss: 0.5564 - val_acc: 0.8539
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.5087 - acc: 0.8550 - val_loss: 0.4721 - val_acc

#selu

keras.activations.selu(x)

Scaled Exponential Linear Unit (SELU).

SELU is equal to: scale * elu(x, alpha), where alpha and scale are predefined constants. The values of alpha and scale are chosen so that the mean and variance of the inputs are preserved between two consecutive layers as long as the weights are initialized correctly (see lecun_normal initialization) and the number of inputs is "large enough" (see references for more information).

##Arguments

1. x: A tensor or variable to compute the activation function for.

##Returns

The scaled exponential unit activation: scale * elu(x, alpha).

##Note

1. To be used together with the initialization "lecun_normal".
2. To be used together with the dropout variant "AlphaDropout".

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='selu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='selu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.3339 - acc: 0.8992 - val_loss: 0.2149 - val_acc: 0.9319
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1353 - acc: 0.9590 - val_loss: 0.1086 - val_acc: 0.9673
Epoch 3/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0938 - acc: 0.9714 - val_loss: 0.0887 - val_acc: 0.9730
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0720 - acc: 0.9772 - val_loss: 0.0953 - val_acc: 0.9693
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0592 - acc: 0.9810 - val_loss: 0.0808 - val_acc: 0.9751
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0486 - acc: 0.9841 - val_loss: 0.0808 - val_acc: 0.9755
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0415 - acc: 0.9861 - val_loss: 0.0862 - val_acc

#softplus

keras.activations.softplus(x)

Softplus activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softplus activation: log(exp(x) + 1).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softplus', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softplus'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.4494 - acc: 0.8619 - val_loss: 0.2046 - val_acc: 0.9345
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1736 - acc: 0.9470 - val_loss: 0.1176 - val_acc: 0.9628
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1119 - acc: 0.9659 - val_loss: 0.1073 - val_acc: 0.9688
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0833 - acc: 0.9737 - val_loss: 0.0896 - val_acc: 0.9730
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0651 - acc: 0.9795 - val_loss: 0.0861 - val_acc: 0.9739
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0521 - acc: 0.9838 - val_loss: 0.0838 - val_acc: 0.9745
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0435 - acc: 0.9862 - val_loss: 0.0778 - val_acc

#softsign

keras.activations.softsign(x)

##Softsign activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softsign activation: x / (abs(x) + 1).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softsign', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softsign'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.2944 - acc: 0.9135 - val_loss: 0.2009 - val_acc: 0.9404
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1445 - acc: 0.9561 - val_loss: 0.1240 - val_acc: 0.9613
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0995 - acc: 0.9694 - val_loss: 0.0943 - val_acc: 0.9720
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0743 - acc: 0.9774 - val_loss: 0.0856 - val_acc: 0.9748
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0576 - acc: 0.9829 - val_loss: 0.0816 - val_acc: 0.9745
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0451 - acc: 0.9861 - val_loss: 0.0749 - val_acc: 0.9755
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0357 - acc: 0.9888 - val_loss: 0.0757 - val_acc

#relu

keras.activations.relu(x, alpha=0.0, max_value=None, threshold=0.0)

##Rectified Linear Unit.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = alpha * (x - threshold) otherwise.

##Arguments

1. x: Input tensor.
2. alpha: float. Slope of the negative part. Defaults to zero.
3. max_value: float. Saturation threshold.
4. threshold: float. Threshold value for thresholded activation.

##Returns

A tensor.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.2273 - acc: 0.9310 - val_loss: 0.0998 - val_acc: 0.9688
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0850 - acc: 0.9739 - val_loss: 0.0734 - val_acc: 0.9770
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0540 - acc: 0.9834 - val_loss: 0.0756 - val_acc: 0.9774
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0396 - acc: 0.9877 - val_loss: 0.0855 - val_acc: 0.9770
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0294 - acc: 0.9909 - val_loss: 0.0770 - val_acc: 0.9783
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0217 - acc: 0.9931 - val_loss: 0.0800 - val_acc: 0.9818
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0173 - acc: 0.9945 - val_loss: 0.0734 - val_acc

#tanh

keras.activations.tanh(x)

##Hyperbolic tangent activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The hyperbolic activation: tanh(x) = (exp(x) - exp(-x)) / (exp(x) + exp(-x))

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='tanh', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='tanh'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.2841 - acc: 0.9144 - val_loss: 0.1427 - val_acc: 0.9564
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1235 - acc: 0.9628 - val_loss: 0.1045 - val_acc: 0.9681
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0815 - acc: 0.9750 - val_loss: 0.0951 - val_acc: 0.9704
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0592 - acc: 0.9819 - val_loss: 0.0830 - val_acc: 0.9730
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0434 - acc: 0.9870 - val_loss: 0.0757 - val_acc: 0.9766
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0317 - acc: 0.9904 - val_loss: 0.0732 - val_acc: 0.9768
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0235 - acc: 0.9930 - val_loss: 0.0725 - val_acc

#sigmoid

keras.activations.sigmoid(x)

##Sigmoid activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The sigmoid activation: 1 / (1 + exp(-x)).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.4923 - acc: 0.8595 - val_loss: 0.2385 - val_acc: 0.9275
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.2153 - acc: 0.9356 - val_loss: 0.1705 - val_acc: 0.9481
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1540 - acc: 0.9532 - val_loss: 0.1375 - val_acc: 0.9590
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1159 - acc: 0.9650 - val_loss: 0.1084 - val_acc: 0.9662
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0909 - acc: 0.9722 - val_loss: 0.0920 - val_acc: 0.9716
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0737 - acc: 0.9777 - val_loss: 0.0852 - val_acc: 0.9740
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0604 - acc: 0.9809 - val_loss: 0.0807 - val_acc

#hard_sigmoid

keras.activations.hard_sigmoid(x)

##Hard sigmoid activation function.

Faster to compute than sigmoid activation.

##Arguments

1. x: Input tensor.

##Returns

1. Hard sigmoid activation:

```
0 if x < -2.5
1 if x > 2.5
0.2 * x + 0.5 if -2.5 <= x <= 2.5.
```

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='hard_sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='hard_sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.5035 - acc: 0.8561 - val_loss: 0.2442 - val_acc: 0.9312
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.2098 - acc: 0.9377 - val_loss: 0.1812 - val_acc: 0.9448
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1464 - acc: 0.9555 - val_loss: 0.1251 - val_acc: 0.9609
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1090 - acc: 0.9679 - val_loss: 0.1087 - val_acc: 0.9685
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0854 - acc: 0.9743 - val_loss: 0.0956 - val_acc: 0.9717
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0685 - acc: 0.9795 - val_loss: 0.0842 - val_acc: 0.9753
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0554 - acc: 0.9833 - val_loss: 0.0891 - val_acc

#exponential

keras.activations.exponential(x)

Exponential (base e) activation function.

##Arguments

1. x: Input tensor.

##Returns

1. Exponential activation: exp(x).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='exponential', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='exponential'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 14.5152 - acc: 0.0984 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.

#linear

keras.activations.linear(x)

Linear (i.e. identity) activation function.

##Arguments

1. x: Input tensor.

##Returns

Input tensor, unchanged.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='linear', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='linear'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.4296 - acc: 0.8761 - val_loss: 0.3429 - val_acc: 0.9017
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.3427 - acc: 0.9025 - val_loss: 0.3263 - val_acc: 0.9066
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.3219 - acc: 0.9094 - val_loss: 0.3062 - val_acc: 0.9147
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.3074 - acc: 0.9131 - val_loss: 0.3242 - val_acc: 0.9099
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.3002 - acc: 0.9156 - val_loss: 0.2907 - val_acc: 0.9205
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.2951 - acc: 0.9168 - val_loss: 0.2933 - val_acc: 0.9198
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.2892 - acc: 0.9189 - val_loss: 0.2924 - val_acc

#LeakyReLU

keras.layers.LeakyReLU(alpha=0.3)

Leaky version of a Rectified Linear Unit.

It allows a small gradient when the unit is not active: 
```
f(x) = alpha * x for x < 0
f(x) = x for x >= 0.
```

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

alpha: float >= 0. Negative slope coefficient.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.LeakyReLU(alpha=0.3), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.LeakyReLU(alpha=0.3)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as LeakyReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.2676 - acc: 0.9182 - val_loss: 0.1356 - val_acc: 0.9603
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1272 - acc: 0.9612 - val_loss: 0.1250 - val_acc: 0.9605
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0929 - acc: 0.9712 - val_loss: 0.0994 - val_acc: 0.9700
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0740 - acc: 0.9773 - val_loss: 0.0925 - val_acc: 0.9728
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0638 - acc: 0.9800 - val_loss: 0.1135 - val_acc: 0.9647
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0547 - acc: 0.9830 - val_loss: 0.0860 - val_acc: 0.9750
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0481 - acc: 0.9849 - val_loss: 0.0994 - val_acc

#PReLU

keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)

Parametric Rectified Linear Unit.

It follows: f(x) = alpha * x for x < 0, f(x) = x for x >= 0, where alpha is a learned array with the same shape as x.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha_initializer: initializer function for the weights.
2. alpha_regularizer: regularizer for the weights.
3. alpha_constraint: constraint for the weights.
4. shared_axes: the axes along which to share learnable parameters for the activation function. For example, if the incoming feature maps are from a 2D convolution with output shape (batch, height, width, channels), and you wish to share parameters across space so that each filter only has one set of parameters, set shared_axes=[1, 2].

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as PReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 42us/step - loss: 0.2256 - acc: 0.9319 - val_loss: 0.1013 - val_acc: 0.9679
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0843 - acc: 0.9738 - val_loss: 0.0766 - val_acc: 0.9780
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0557 - acc: 0.9828 - val_loss: 0.0683 - val_acc: 0.9809
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0393 - acc: 0.9876 - val_loss: 0.0848 - val_acc: 0.9777
Epoch 5/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0291 - acc: 0.9908 - val_loss: 0.0822 - val_acc: 0.9769
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0223 - acc: 0.9932 - val_loss: 0.0828 - val_acc: 0.9809
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0180 - acc: 0.9943 - val_loss: 0.0861 - val_acc

#ELU

keras.layers.ELU(alpha=1.0)

Exponential Linear Unit.

It follows: f(x) =  alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha: scale for the negative factor.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ELU(alpha=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ELU(alpha=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ELU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 39us/step - loss: 0.2918 - acc: 0.9116 - val_loss: 0.1373 - val_acc: 0.9585
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1221 - acc: 0.9625 - val_loss: 0.1112 - val_acc: 0.9645
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0832 - acc: 0.9746 - val_loss: 0.0935 - val_acc: 0.9699
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0632 - acc: 0.9800 - val_loss: 0.0810 - val_acc: 0.9754
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0505 - acc: 0.9840 - val_loss: 0.0869 - val_acc: 0.9725
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0409 - acc: 0.9867 - val_loss: 0.0956 - val_acc: 0.9735
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0335 - acc: 0.9892 - val_loss: 0.0757 - val_acc

#ThresholdedReLU

keras.layers.ThresholdedReLU(theta=1.0)

Thresholded Rectified Linear Unit.

It follows: f(x) = x for x > theta, f(x) = 0 otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. theta: float >= 0. Threshold location of activation.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ThresholdedReLU(theta=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ThresholdedReLU(theta=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ThresholdedReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.5300 - acc: 0.8282 - val_loss: 0.1811 - val_acc: 0.9435
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1445 - acc: 0.9572 - val_loss: 0.1260 - val_acc: 0.9606
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0979 - acc: 0.9705 - val_loss: 0.1333 - val_acc: 0.9594
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0754 - acc: 0.9769 - val_loss: 0.1080 - val_acc: 0.9686
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0598 - acc: 0.9818 - val_loss: 0.0861 - val_acc: 0.9749
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0490 - acc: 0.9850 - val_loss: 0.0814 - val_acc: 0.9783
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0410 - acc: 0.9875 - val_loss: 0.0754 - val_acc

#Softmax

keras.layers.Softmax(axis=-1)

Softmax activation function.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. axis: Integer, axis along which the softmax normalization is applied.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.Softmax(axis=-1), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.Softmax(axis=-1)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as Softmax) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 43us/step - loss: 2.2692 - acc: 0.1798 - val_loss: 2.1970 - val_acc: 0.6770
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.0416 - acc: 0.8212 - val_loss: 1.8493 - val_acc: 0.8815
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.6269 - acc: 0.9013 - val_loss: 1.3929 - val_acc: 0.9135
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.1774 - acc: 0.9164 - val_loss: 0.9659 - val_acc: 0.9210
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.7995 - acc: 0.9228 - val_loss: 0.6472 - val_acc: 0.9248
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.5441 - acc: 0.9257 - val_loss: 0.4570 - val_acc: 0.9273
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.4040 - acc: 0.9286 - val_loss: 0.3627 - val_acc

#ReLU

keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)

Rectified Linear Unit activation function.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = negative_slope * (x - threshold) otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. max_value: float >= 0. Maximum activation value.
2. negative_slope: float >= 0. Negative slope coefficient.
3. threshold: float. Threshold value for thresholded activation.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.2252 - acc: 0.9315 - val_loss: 0.1103 - val_acc: 0.9672
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0822 - acc: 0.9747 - val_loss: 0.0826 - val_acc: 0.9739
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0545 - acc: 0.9827 - val_loss: 0.0948 - val_acc: 0.9718
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0389 - acc: 0.9879 - val_loss: 0.0732 - val_acc: 0.9782
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0290 - acc: 0.9911 - val_loss: 0.0891 - val_acc: 0.9766
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0220 - acc: 0.9931 - val_loss: 0.0734 - val_acc: 0.9819
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0179 - acc: 0.9945 - val_loss: 0.0821 - val_acc

#Thank you for completing this notebook

**relu** 

Test loss: 0.08459654634964409

Test accuracy: 0.9799

**softmax**

Test loss: 0.39845509810447693

Test accuracy: 0.8642

**selu**

Test loss: 0.11848029121595427

Test accuracy: 0.9726

**softplus**

Test loss: 0.08013557475531798

Test accuracy: 0.9798

**softsign**

Test loss: 0.07156896027913608

Test accuracy: 0.9793

**relu**

Test loss: 0.09273777272929686

Test accuracy: 0.9843

**tanh**

Test loss: 0.06885148700817954

Test accuracy: 0.9803

**sigmoid**

Test loss: 0.07136517209208104

Test accuracy: 0.9781

**hard_sigmoid**

Test loss: 0.06968761095895898

Test accuracy: 0.9795

**exponential**

Test loss: 14.573981648254394

Test accuracy: 0.0958

**linear**

Test loss: 0.30165807257443666

Test accuracy: 0.9158

**LeakyReLU**

Test loss: 0.11553157684192265

Test accuracy: 0.9709

**PReLU**

Test loss: 0.11553157684192265

Test accuracy: 0.9709

**ELU**

Test loss: 0.1115254690002792

Test accuracy: 0.9743

**ThresholdedReLU**

Test loss: 0.08483904730657359

Test accuracy: 0.9786

**Softmax**

Test loss: 0.2818538703083992

Test accuracy: 0.9333

**ReLU**

Test loss: 0.09205757192509677

Test accuracy: 0.9829


The best one is **ReLU** among all the activations 

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.layers import Activation, Flatten, Conv2D

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1))
X_test = X_test.astype('float32') / 255



# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)

# Build neural network
model = models.Sequential()
model.add(Conv2D(512, (28, 28), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 138us/step - loss: 0.5952 - acc: 0.7774 - val_loss: 0.4582 - val_acc: 0.8405
Epoch 2/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.3959 - acc: 0.8543 - val_loss: 0.4307 - val_acc: 0.8490
Epoch 3/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.3511 - acc: 0.8709 - val_loss: 0.3689 - val_acc: 0.8703
Epoch 4/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.3268 - acc: 0.8806 - val_loss: 0.3826 - val_acc: 0.8659
Epoch 5/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.3104 - acc: 0.8866 - val_loss: 0.4175 - val_acc: 0.8533
Epoch 6/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.2976 - acc: 0.8906 - val_loss: 0.3873 - val_acc: 0.8679
Epoch 7/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.2883 - acc: 0.8938 - val_loss: 0.4067 - val_ac

**FASHION_MNIST**

Test loss: 0.4138208844304085

Test accuracy: 0.8699

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar10
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

170500096/170498071 [==============================] - 2s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 5s 96us/step - loss: 2.0376 - acc: 0.2890 - val_loss: 1.7873 - val_acc: 0.3573
Epoch 2/10
50000/50000 [==============================] - 4s 70us/step - loss: 1.7179 - acc: 0.3877 - val_loss: 1.6877 - val_acc: 0.4077
Epoch 3/10
50000/50000 [==============================] - 4s 71us/step - loss: 1.6274 - acc: 0.4201 - val_loss: 1.6767 - val_acc: 0.4093
Epoch 4/10
50000/50000 [==============================] - 3s 70us/step - loss: 1.5713 - acc: 0.4439 - val_loss: 1.5704 - val_acc: 0.4346
Epoch 5/10
50000/50000 [==============================] - 3s 70us/step - loss: 1.5193 - acc: 0.4598 - val_loss: 1.5190 - val_acc: 0.4570
Epoch 6/10
50000/50000 [==============================] - 3s 69us/step - loss: 1.4744 - acc: 0.4746 - val_loss: 1.4632 - val_acc: 0.4809
Epoch 7/10
50000/50000 [==============================] - 3s 68us/step - loss: 1.4309 - acc: 0.4923 - val_loss: 1.4382 - val_acc

**CIFAR10**

Test loss: 1.4009930366516112

Test accuracy: 0.4971


In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar100
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 100)
Y_test =  to_categorical(y_test, 100)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(100, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 2s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 5s 95us/step - loss: 4.3099 - acc: 0.0508 - val_loss: 4.0004 - val_acc: 0.0884
Epoch 2/10
50000/50000 [==============================] - 4s 70us/step - loss: 3.8491 - acc: 0.1114 - val_loss: 3.7959 - val_acc: 0.1133
Epoch 3/10
50000/50000 [==============================] - 4s 72us/step - loss: 3.6442 - acc: 0.1473 - val_loss: 3.5849 - val_acc: 0.1562
Epoch 4/10
50000/50000 [==============================] - 4s 72us/step - loss: 3.5108 - acc: 0.1681 - val_loss: 3.5200 - val_acc: 0.1725
Epoch 5/10
50000/50000 [==============================] - 4s 73us/step - loss: 3.3917 - acc: 0.1911 - val_loss: 3.4822 - val_acc: 0.1804
Epoch 6/10
50000/50000 [==============================] - 4s 71us/step - loss: 3.2977 - acc: 0.2070 - val_loss: 3.3675 - val_acc: 0.2014
Epoch 7/10
50000/50000 [==============================] - 4s 72us/step - loss: 3.2009 - acc: 0.2235 - val_loss: 3.2949 - val_acc

**CIFAR100**

Test loss: 3.242277897644043

Test accuracy: 0.2326